emu/src/bus.rs

```diff
@@ -32,12 +32,18 @@ impl Bus {
     }
 
     fn step(&mut self) {
+        // Step cycles at beginning or end?
+        // It may have an impact when we will introduce timers.
         self.cycles_count += 1;
 
         // TODO: move this somewhere in the UI
         log(format!("CPU Cycles: {}", self.cycles_count));
 
         // Step ppu, dma, interrupts, timers, etc...
+        let mut mem = self.internal_memory.lock().unwrap();
+
+        let val = *mem.interrupts.interrupt_request.back().unwrap();
+        mem.interrupts.interrupt_request.push(val);
```

Comment:
I don't fully understand this, removing from back and pushing to front... 🤔

Reply:
With the first line I don't want to remove from the back, the behaviour I'm trying to have is:

`interrupt_request = [1, 0, 0, 0, 3]`

I execute these two lines and I want:

`interrupt_request = [1, 1, 0, 0, 0]`

Reply:
So I want to take the back and push a copy of it

Reply:
The reason is that once I write to the interrupt requests register, it is delayed by 4 cycles before the CPU sees it.

So my idea was to write the back when we write the memory and the CPU sees the front of this buffer.
Every step we propagate the information so:

`T = 0`: `interrupt_request = [0, 0, 0, 0, 1]` (requesting an interrupt like hblank, vblank or something like that)
`T = 1`: `interrupt_request = [0, 0, 0, 1, 1]`
`T = 2`: `interrupt_request = [0, 0, 1, 1, 1]` 
`T = 3`: `interrupt_request = [0, 1, 1, 1, 1]`
`T = 4`: `interrupt_request = [1, 1, 1, 1, 1]`  

So the CPU sees the `1` at `T=4`

Reply:
Mmm ok, got it thanks...